R/cli.R

```diff
@@ -102,17 +102,24 @@ cli_fmt <- function(expr, collapse = FALSE, strip_newline = FALSE) {
 #' @param .envir Environment to evaluate the expressions in.
 #' @param collapse Whether to collapse the result if it has multiple
 #'   lines, e.g. because of `\f` characters.
+#' @param keep_newlines Whether to keep newlines in the result, or treat
+#'   them as regular space characters.
 #' @return Character scalar, the formatted string.
 #'
 #' @export
 #' @examples
 #' format_inline("A message for {.emph later}, thanks {.fn format_inline}.")
 
-format_inline <- function(..., .envir = parent.frame(), collapse = TRUE) {
+format_inline <- function(..., .envir = parent.frame(), collapse = TRUE,
+                          keep_newlines = TRUE) {
+  str <- paste0(unlist(list(...), use.names = FALSE), collapse = "")
+  if (keep_newlines) {
+    str <- gsub("\n", "\f", str, fixed = TRUE)
```

Comment:
IIUC this means the newlines are set in stone and can no longer be reformatted on subsequent calls to e.g. `format_error()`? I.e. when rlang calls back cli at display time?

I was thinking `format_inline()` should preserve newlines but should not affect how newlines are interpreted by other cli functions.

Reply:
ah no with dev cli I see that form feeds are transformed to newlines:

```r
cli::format_inline("foo\fbar\nbaz")
#> [1] "foo\nbar\nbaz"
```

But then this means that form feeds are not preserved by `format_inline()`, which I think explains why they keep being reformatted by `cli::cli_abort()` since rlang calls cli back with a `\f` transformed to a `\n`?

Reply:
cli treats form feeds as paragraph separators and newlines as spaces. This is not new. What is the behaviour you want for rlang?

Reply:
The behaviour outlined in: https://github.com/r-lib/rlang/issues/1535#issuecomment-1373646830

- `format_inline()` is called by `cli_abort()` at throw time to interpolate and apply style selections.
- `format_error()` is called by rlang at display time to format the layout of the message (width wrapping, bullets).

So `format_inline()` should not have an effect on the eventual layout formatting.

Note that this is not only for rlang or `cli::cli_abort()`. When we assemble messages from pieces in cnd-message methods, we often call `format_inline()` instead of `glue()` to interpolate and apply styles. Then at the end the message is formatted using `format_error()`. In these cases `format_inline()` should preserve both `\n` and `\f` as well so that they can be interpreted by the final message/layout formatter.

Reply:
OK, I changed `format_inline()` to keep all whitespace as is.

Reply:
Thanks!
I just tried it and it solves the `\f` handling in `cli_abort()` :)

```r
cli::cli_abort("foo\fbar")
#> Error:
#> ! foo
#> bar
```